pkg/drift/driftignore.go

```diff
@@ -0,0 +1,143 @@
+
+// filterIgnored removes any asset iam that is in the ignored assets.
+func filterIgnored(values map[string]*iam.AssetIAM, ignored *ignoredAssets) map[string]*iam.AssetIAM {
+	filtered := make(map[string]*iam.AssetIAM)
+	for k, a := range values {
+		if a.ResourceType == assets.Project {
+			if _, ok := ignored.projects[a.ResourceID]; !ok {
+				filtered[k] = a
+			}
+		} else if a.ResourceType == assets.Folder {
+			if _, ok := ignored.folders[a.ResourceID]; !ok {
+				filtered[k] = a
+			}
+		} else { // Handle default case so we do not accidentally drop values.
+			filtered[k] = a
+		}
+	}
+
+	return filtered
+}
+
+// expandGraph traverses the asset hierarchy graph and adds any nested folders or projects beneath every ignored asset.
+func expandGraph(ignored *ignoredAssets, hierarchyGraph *assets.HierarchyGraph) (*ignoredAssets, error) {
+	ignoredProjects := ignored.projects
+	ignoredFolders := ignored.folders
+
+	// Traverse the hierarchy
+	for folderID := range ignored.folders {
+		ids, err := assets.FoldersBeneath(folderID, hierarchyGraph)
+		if err != nil {
+			return nil, fmt.Errorf("failed to traverse hierarchy for folder with ID %s: %w", folderID, err)
+		}
+		ignoredFolders = mergeSets(ignoredFolders, ids)
+	}
+	for folderID := range ignoredFolders {
+		ignoredProjects = addListToSet(ignoredProjects, hierarchyGraph.IDToNodes[folderID].ProjectIDs)
+	}
+
+	return &ignoredAssets{
+		iamAssets: ignored.iamAssets,
+		projects:  ignoredProjects,
+		folders:   ignoredFolders,
+	}, nil
+}
+
+// driftignore parses the driftignore file into a set.
+// Go doesn't implement set so we use a map of boolean values all set to true.
+func driftignore(ctx context.Context, fname string) (*ignoredAssets, error) {
+	iamAssets := make(map[string]struct{})
+	projects := make(map[string]struct{})
+	folders := make(map[string]struct{})
+	f, err := os.Open(fname)
+	if err != nil {
+		if os.IsNotExist(err) {
+			logger := logging.FromContext(ctx)
+			logger.Debugw("failed to find driftignore", "filename", fname)
+			return &ignoredAssets{
+				iamAssets,
+				projects,
+				folders,
+			}, nil
+		}
+		return nil, fmt.Errorf("failed to read driftignore file %s: %w", fname, err)
+	}
+	defer f.Close()
+
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+		iamAssets[line] = struct{}{}
+
+		projectMatches := ignoredProjectPattern.FindStringSubmatch(line)
+		if len(projectMatches) == 2 {
+			projects[projectMatches[1]] = struct{}{}
+		}
+
+		folderMatches := ignoredFolderPattern.FindStringSubmatch(line)
+		if len(folderMatches) == 2 {
+			folders[folderMatches[1]] = struct{}{}
+		}
+	}
+
+	return &ignoredAssets{
+		iamAssets,
+		projects,
+		folders,
+	}, nil
+}
+
+func addListToSet(set map[string]struct{}, list []string) map[string]struct{} {
+	for _, v := range list {
+		set[v] = struct{}{}
+	}
+	return set
+}
+
+func mergeSets(setA, setB map[string]struct{}) map[string]struct{} {
```

Comment:
Can we add a doc comment here saying that setA is mutated? The caller might be surprised that setA is an output parameter.

Reply:
Or even better, remove the return value. That will make it more clear that the argument is mutated in place.

Reply:
Done